primitives/src/merkle_tree/namespaced_merkle_tree.rs

```diff
@@ -181,10 +188,387 @@ where
+    }
+}
+
+impl<E, H, Arity, N, T> AppendableMerkleTreeScheme for NMT<E, H, Arity, N, T>
+where
+    H: DigestAlgorithm<E, u64, T> + BindNamespace<E, u64, T, N>,
+    E: Element + Namespaced<Namespace = N>,
+    T: NodeValue,
+    N: Namespace,
+    Arity: Unsigned,
+{
+    fn extend(
+        &mut self,
+        elems: impl IntoIterator<Item = impl core::borrow::Borrow<Self::Element>>,
+    ) -> Result<(), PrimitivesError> {
+        // TODO: update namespace metadata
+        self.inner.extend(elems)
+    }
+
+    fn push(
+        &mut self,
+        elem: impl core::borrow::Borrow<Self::Element>,
+    ) -> Result<(), PrimitivesError> {
+        // TODO: update namespace metadata
+        self.inner.push(elem)
+    }
+}
+
+#[derive(Clone, Debug)]
+enum NamespaceProofType {
+    Presence,
+    Absence,
+}
+
+#[derive(Clone, Debug)]
+/// Namespace Proof
+pub struct NamespaceProof<E, T, Arity, N, H>
+where
+    E: Element + Namespaced<Namespace = N>,
+    T: NodeValue,
+    H: DigestAlgorithm<E, u64, T> + BindNamespace<E, u64, T, N>,
+    N: Namespace,
+    Arity: Unsigned,
+{
+    proof_type: NamespaceProofType,
+    proofs: Vec<MerkleProof<E, u64, NamespacedHash<T, N>, Arity>>,
+    left_boundary_proof: Option<MerkleProof<E, u64, NamespacedHash<T, N>, Arity>>,
+    right_boundary_proof: Option<MerkleProof<E, u64, NamespacedHash<T, N>, Arity>>,
+    indices: Vec<u64>,
+    phantom: PhantomData<H>,
+}
+impl<E, T, Arity, N, H> NamespaceProof<E, T, Arity, N, H>
+where
+    E: Element + Namespaced<Namespace = N>,
+    T: NodeValue,
+    H: DigestAlgorithm<E, u64, T> + BindNamespace<E, u64, T, N>,
+    N: Namespace,
+    Arity: Unsigned,
+{
+    /// Return the set of leaves associated with this Namespace proof as an
+    /// iterator
+    pub fn get_namespace_leaves(&self) -> impl Iterator<Item = &E> {
+        let num_leaves = match self.proof_type {
+            NamespaceProofType::Presence => self.proofs.len(),
+            NamespaceProofType::Absence => 0,
+        };
+        self.proofs
+            .iter()
+            .map(|proof| proof.elem().unwrap())
+            .take(num_leaves)
+    }
+}
+
+impl<E, T, Arity, N, H> NamespaceProof<E, T, Arity, N, H>
+where
+    E: Element + Namespaced<Namespace = N>,
+    T: NodeValue,
+    H: DigestAlgorithm<E, u64, T> + BindNamespace<E, u64, T, N>,
+    N: Namespace,
+    Arity: Unsigned,
+{
+    fn verify_left_namespace_boundary(
+        &self,
+        root: &NamespacedHash<T, N>,
+        namespace: N,
+    ) -> Result<VerificationResult, PrimitivesError> {
+        if let Some(boundary_proof) = self.left_boundary_proof.as_ref() {
+            // If there is a leaf to the left of the namespace range, check that it is less
+            // than the target namespace
+            if boundary_proof.elem().unwrap().get_namespace() >= namespace
+                || *boundary_proof.index() != self.indices[0] - 1
+            {
+                return Ok(Err(()));
```

Comment:
This is confusing. Why do we need to wrap the result into another result?

Reply:
Actually, I think we can directly return `VerificationResult` here. @ggutoski correct me if I'm wrong, but I believe the nested pattern is used to distinguish user input error (incorrect serialization, incorrect parameters, etc.) from the actual proof verification result.

Reply:
I agree it looks confusing at first glance but there is a reason for it. See discussion https://github.com/EspressoSystems/jellyfish/pull/274#discussion_r1206866522 and rustdoc:

https://github.com/EspressoSystems/jellyfish/blob/466a7604f00a6d5b142ae1b3b7aabcd1111f06df/primitives/src/errors.rs#L18-L23

The nested result pattern is well-described here: https://sled.rs/errors

I've been slowly migrating code to conform to this pattern, though no explicit declaration has been made to make it a policy for jellyfish code. We can discuss further if you're interested.

In this particular case, the method is infallible so technically we can return `VerificationResult` directly as @nomaxg suggests. Is this infallibility an explicit promise we wish to make forever? If not then perhaps we should keep the nested result return type for future-proofness.

Reply:
Since this is a helper function, I think its ok to return `VerificationResult` directly. However, the outer `verify` will need the nested result since it makes a call to the internal merkle `verify` fn.

Personally, I'm a fan of this pattern because it gives a nice separation between different classes of errors.

Reply:
Thank @ggutoski @nomaxg , I was not aware of this error handling pattern. Makes sense.